src/app/clusters/resource-monitoring-server/resource-monitoring-cluster-objects.h

```diff
@@ -141,37 +143,72 @@ struct TypeInfo
+// from the application.
+//
+// Inherit from an auto-generated struct to pick up the implementation bits, but make
+// it private inheritance so people can't accidentally use this struct where the other
+// is expected.
+struct GenericType : private HepaFilterMonitoring::Structs::ReplacementProductStruct::Type
 {
-    kProductIdentifierType  = 0,
-    kProductIdentifierValue = 1,
-};
+private:
+    char productIdentifierValueBuffer[kProductIdentifierValueMaxNameLength];
 
-struct Type
-{
 public:
-    ProductIdentifierTypeEnum productIdentifierType = static_cast<ProductIdentifierTypeEnum>(0);
-    chip::CharSpan productIdentifierValue;
-
-    CHIP_ERROR Decode(TLV::TLVReader & reader);
-
     static constexpr bool kIsFabricScoped = false;
-
-    CHIP_ERROR Encode(TLV::TLVWriter & writer, TLV::Tag tag) const;
+    virtual ~GenericType()                = default;
+    GenericType() {}
+    GenericType(ResourceMonitoring::ProductIdentifierTypeEnum aProductIdentifierType, chip::CharSpan aProductIdentifierValue)
+    {
+        setProductIdentifierType(aProductIdentifierType);
+        setProductIdentifierValue(aProductIdentifierValue);
+    }
+
+    using HepaFilterMonitoring::Structs::ReplacementProductStruct::Type::Encode;
+
+    /**
+     * Sets the product identifier type.
+     *
+     * @param aProductIdentifierType The product identifier type.
+     */
+    void setProductIdentifierType(ResourceMonitoring::ProductIdentifierTypeEnum aProductIdentifierType)
+    {
+        productIdentifierType = static_cast<HepaFilterMonitoring::ProductIdentifierTypeEnum>(aProductIdentifierType);
+    }
+
+    /**
+     * Sets the product identifier value.
+     * This implementation will copy the argument to the local implementation.
+     *
+     * @param aProductIdentifierValue The value of the product identifier to set.
+     * @return CHIP_ERROR_INVALID_ARGUMENT when aProductIdentifierValue is invalid, CHIP_NO_ERROR otherwise.
+     */
+    CHIP_ERROR setProductIdentifierValue(chip::CharSpan aProductIdentifierValue)
+    {
+        VerifyOrReturnError(IsSpanUsable(aProductIdentifierValue), CHIP_ERROR_INVALID_ARGUMENT);
+
+        if (aProductIdentifierValue.size() > sizeof(productIdentifierValueBuffer))
+        {
+            memcpy(productIdentifierValueBuffer, aProductIdentifierValue.data(), sizeof(productIdentifierValueBuffer));
+            productIdentifierValue = CharSpan(productIdentifierValueBuffer, sizeof(productIdentifierValueBuffer));
+        }
+        else
+        {
+            memcpy(productIdentifierValueBuffer, aProductIdentifierValue.data(), aProductIdentifierValue.size());
+            productIdentifierValue = CharSpan(productIdentifierValueBuffer, aProductIdentifierValue.size());
+        }
+
+        return CHIP_NO_ERROR;
+    }
 };
 
-using DecodableType = Type;
-
-} // namespace ReplacementProductStruct
-
 namespace ReplacementProductList {
 static constexpr AttributeId Id = 0x00000005;
 struct TypeInfo
```

Comment:
I am very certain you do not need this TypeInfo thing at all, actually.  It's only used on the client side.